### PR TITLE
Use quaternions in yzy to zyz transformation

### LINKED ISSUE #520: Failed to compile Grover's circuit on ibmqx5
qiskit failed to compile a circuit for Grover algorithms as described in this tutorial:

https://github.com/rraymondhp/qiskit-tutorial/blob/37b3229c8988bbaf1edda744de456cf195e25fc7/reference/algorithms/grover_algorithm.ipynb

## Expected Behavior

qiskit v0.4.15 could compile the same program on ibmqx5 coupling map. 

## Current Behavior

AssertionError that should not happen, happened. 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-9-93ef2321f491> in <module>()
     30 backend_coupling = backend_config['coupling_map']
     31 
---> 32 compiled_circuit = qp.compile([circuitName], backend=backend, coupling_map=backend_coupling)
     33 compiled_qasm = qp.get_compiled_qasm(compiled_circuit, circuitName)
     34 print("Number of gates at ", circuitName, " circuit is ", len(compiled_qasm.split("\n")) - 4)

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/_quantumprogram.py in compile(self, name_of_circuits, backend, config, basis_gates, coupling_map, initial_layout, shots, max_credits, seed, qobj_id, hpc, skip_translation)
    972                                       config, basis_gates, coupling_map, initial_layout,
    973                                       shots, max_credits, seed, qobj_id, hpc,
--> 974                                       skip_translation)
    975         return qobj
    976 

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/wrapper/_wrapper.py in compile(circuits, backend, config, basis_gates, coupling_map, initial_layout, shots, max_credits, seed, qobj_id, hpc, skip_translation)
    171                                     config, basis_gates, coupling_map, initial_layout,
    172                                     shots, max_credits, seed, qobj_id, hpc,
--> 173                                     skip_translation)
    174 
    175 

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/_compiler.py in compile(circuits, backend, config, basis_gates, coupling_map, initial_layout, shots, max_credits, seed, qobj_id, hpc, skip_translation)
    146                 coupling_map=coupling_map,
    147                 initial_layout=initial_layout,
--> 148                 get_layout=True)
    149             # Map the layout to a format that can be json encoded
    150             list_layout = None

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/_compiler.py in compile_circuit(quantum_circuit, basis_gates, coupling_map, initial_layout, get_layout, format, seed)
    237         cx_cancellation(compiled_dag_circuit)
    238         # Simplify single qubit gates
--> 239         compiled_dag_circuit = optimize_1q_gates(compiled_dag_circuit)
    240         logger.info("post-mapping properties: %s",
    241                     compiled_dag_circuit.property_summary())

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/mapper/_mapping.py in optimize_1q_gates(circuit)
    849                                               right_parameters[0],
    850                                               right_parameters[1],
--> 851                                               right_parameters[2])
    852                 # Why evalf()? This program:
    853                 #   OPENQASM 2.0;

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/mapper/_mapping.py in compose_u3(theta1, phi1, lambda1, theta2, phi2, lambda2)
    736     # Careful with the factor of two in yzy_to_zyz
    737     thetap, phip, lambdap = yzy_to_zyz((lambda1 + phi2) / 2,
--> 738                                        theta1 / 2, theta2 / 2)
    739     (theta, phi, lamb) = (2 * thetap, phi1 + 2 * phip, lambda2 + 2 * lambdap)
    740     return (theta.simplify(), phi.simplify(), lamb.simplify())

~/miniconda3/envs/QISKitenv/lib/python3.6/site-packages/qiskit/mapper/_mapping.py in yzy_to_zyz(xi, theta1, theta2, eps)
    720     logger.debug("solutions=%s", pprint.pformat(solutions))
    721     logger.debug("deltas=%s", pprint.pformat(deltas))
--> 722     assert False, "Error! No solution found. This should not happen."
    723 
    724 

AssertionError: Error! No solution found. This should not happen.

```

## Steps to Reproduce (for bugs)

1. Run this tutorial: https://github.com/rraymondhp/qiskit-tutorial/blob/37b3229c8988bbaf1edda744de456cf195e25fc7/reference/algorithms/grover_algorithm.ipynb
2.
3.
4.

## Your Environment

* Version used: qiskit-sdk-py version 0.5.3
* Environment name and version (e.g. Python 3.6.1): Python 3.6.x
* Operating System and version: MacOSX

### LINKED ISSUE #607: mapping from  Euler angles to quaternions in mapper
### What is the expected enhancement?
To fix some of the mapper i going to change the 1q optimization from euler angles to quaternions and it should fix the mapper issues that keep getting brought up. 

This is a low priority as most times it works but the correct way to do it is quaternions.

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
This fixes the errors in the `qiskit.mapper._mapping.yzy_to_zyz` Euler angle mapping by switching to quaternions.  

### Details and comments
The previous routine was failing for several instances.  I believe the code itself had some math errors that made the transformations invalid rather than we were running into issues with the use of Euler angles themselves.  Regardless, this is an implementation that computes the `yzy` rotation using quaternions, and returns the `zyz` from the generated rotation matrix.  The routine checks the validity of the transformation by computing the quaternions in both `yzy` and `zyz` and checks to see if the abs inner product is close to one.  Additional sanity tests are also implemented.

The failed examples in #607 now generate results.

Fixes #520, Fixes #607.

### CODE REVIEW

**Comment by ajavadia** [comment]
awesome. can you add the circuits from #607 as tests in `test/python/test_mapping.py`? There are some other tests in there already about previously-raised mapper issues. We should ensure these issues are resolved and remain resolved in the future.

**Comment by nonhermitian** [comment]
Do you want to just verify that the circuit compile fine, or do you want the output to be equated with some known result?  They compile just fine now.

**Comment by ajavadia** [comment]
It should be a unit test on compilation, verifying that the compiled output qobj is correctly generated.
It shouldn't include execution as part of the test.

**Comment by diego-plan9** [comment]
Great! Can we also add a note in the `Fixed` section of the changelog, as it is a relevant fix?

**Comment by jaygambetta** [comment]
looks great @nonhermitian, i agree with @ajavadia that we should use those circuits as part of the test. 

@rraymondhp the fix will be in place soon :-)

**Comment by nonhermitian** [comment]
Confirmed that #520 works now as well.

**Comment by rraymondhp** [comment]
I am amazed with the speed of doing this! I really appreciate it.

File: test/python/test_mapper.py [hunk 1]
```diff
@@ -200,6 +199,26 @@ def test_already_mapped(self):
...
         self.assertEqual(sorted(cx_qubits), [[3, 4], [3, 14], [5, 4], [9, 8], [12, 11], [13, 4]])
 
+    @requires_qe_access
+    def test_yzy_zyz_cases(self, QE_TOKEN, QE_URL, hub, group, project):
+        """Test mapper function yzy_to_zyz works in previously failed cases.
+
+        See: https://github.com/QISKit/qiskit-terra/issues/607
+        """
+        qiskit.wrapper.register(QE_TOKEN, QE_URL, hub, group, project)
+        if 'ibmqx4' in qiskit.wrapper.available_backends():
```
Comment (ajavadia) [inline test/python/test_mapper.py]: why do you need to hardcode the backend here? also this is for compilation not execution, so i think just using `compile()` and passing your own `coupling_map` should work.

File: test/python/test_mapper.py [hunk 2]
```diff
@@ -200,6 +199,26 @@ def test_already_mapped(self):
...
+        elif 'ibmq_20_tokyo' in qiskit.wrapper.available_backends():
+            backend = 'ibmq_20_tokyo'
+        else:
+            raise unittest.SkipTest('Do not have the required backend.')
+        circ1 = load_qasm_string(yzy_zyz_1)
+        qobj1 = qiskit.wrapper.compile(circ1, backend)
+        self.assertIsInstance(qobj1, dict)
+        circ2 = load_qasm_string(yzy_zyz_2)
+        qobj2 = qiskit.wrapper.compile(circ2, backend)
+        self.assertIsInstance(qobj2, dict)
```
Comment (ajavadia) [inline test/python/test_mapper.py]: In addition to testing that qobj is created, is it possible to make sure the correct qobj was created? e.g. check one of the rotation angles against expected.

Comment (nonhermitian) [reply test/python/test_mapper.py]: In some sense this is what the `abs_inner` check is doing.  Given an input sequence of rotations, it guarantees that the output sequence gives the same rotation matrix to a tolerance set by the `eps` parameter.  In this sense, we are checking all the time for validity, which is why that condition is in there.

File: test/python/test_mapper.py (hunk 1 above)
Comment (nonhermitian) [reply test/python/test_mapper.py]: I made a fake QX4 backend to test against.

File: test/python/test_mapper.py (hunk 2 above)
Comment (nonhermitian) [reply test/python/test_mapper.py]: Here, we only do `yzy` to `zyz` transformations.  However, the tests check for differences between any valid input Euler angles and the output `zyz` angles, so as to be confident that we are actually doing the right transform regardless of initial basis choice.

**Comment by ajavadia** [comment]
Thanks for crossing an old item off the todo list!

**Pull request merged**